### PR TITLE
Edge walls and End

### PR DESCRIPTION
**Pull request by jorisBarkema**
There are always walls on the edges now, and end tiles will only be placed on the edge of the maze.

There are still some issues, I'm just makign this a PR so I can get reviews.

UpdateGraphicsToMatchSurroundings is not called in the level editor wchich does start with walls now on the edges, graphical bug.

I tested the adapted mazegenerator a few times, exit seems to be working correctly but I'd like some input on this.

The Tile.Add function is kind of rough now, ideally it should show some kind of error message if you try to place a tile somewhere you can't, but I don't now how to do that.

### CODE REVIEW

**Review by jensj12 — CHANGES_REQUESTED** [review]
Remove all changes from the maze generator, only conflicts. Make the endTile a subclass of walltile instead of checking for endtiles in ugtms(). Make the rotation of the exit depend on the location, not surroundings.

**Comment by jorisBarkema** [comment]
Do you mean merge conflicts? Because it doesn't say there are any for me and we are going to have to change it to only make exits on the edges.

**Comment by maxmeijer42** [comment]
It's not conflicted with the master but with #135

**Comment by jorisBarkema** [comment]
aaahhh  I get it alright I'll leave it out for now, save it somewhere so we can re-add it once #135 is in master.

File: MeesGame/Gameplay/GameObjects/TileField.cs [hunk 1]
```diff
@@ -53,6 +53,31 @@ public TileType GetType(int x, int y)
 
         public void Add(Tile obj, int x, int y)
         {
+            // If You are on the edge, but not on a corner, you can place and EndTile
+            if (OnEdgeOfTileField(x, y) && !OnCornerOfTileField(x, y) && obj.TileType == TileType.End)
+            {
+                base.Add(obj, x, y);
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/TileField.cs]: Please rework this section. Do not change the tile if the selected tile isn't allowed. Don't use more than one base.Add(). Simply return if the tile isn't allowed.

File: MeesGame/Gameplay/GameObjects/TileField.cs [hunk 2]
```diff
@@ -85,7 +110,7 @@ public Tile GetTile(int x, int y)
             return (Tile)Get(x, y);
         }
 
-        ///this bool returns if the x and y are outside the bounds of the tilefield
+        /// This bool returns if the x and y are outside the bounds of the tilefield.
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/TileField.cs]: Summary tags are missing

**Review by jensj12 — CHANGES_REQUESTED** [review]

**Comment by jorisBarkema** [comment]
I see the maze generator refactor is on master, is it going to stay this way for a while now? Because in that case I'll update this branch from master and implement my changes.

**Comment by jensj12** [comment]
If you can get your changes in before someone refactors it again (which probably won't happen again), you should implement it.

**Comment by jorisBarkema** [comment]
Alright I implemented it.

**Comment by jorisBarkema** [comment]
There's an error when starting the game with the UGTMS in tilefield without parameter, nullreferenceException unhandled, but I have no idea where it comes from.

File: MeesGame/Gameplay/GameObjects/TileField.cs [hunk 3]
```diff
@@ -239,10 +293,10 @@ public Vector2 CellDimensions
             }
         }
 
-        /// <summary>
-        /// Should be called when the player enters a Tile so that the graphics can be adjusted.
-        /// </summary>
-        /// <param name="location">The location of the Tile the player visits</param>
+        /// <summary>		
+        /// Should be called when the player enters a Tile so that the graphics can be adjusted.		
+        /// </summary>
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/TileField.cs]: Don't add extra whitespace.

File: MeesGame/Gameplay/GameObjects/TilesTypes/EndTile.cs [hunk 4]
```diff
@@ -10,5 +10,21 @@ public override void EnterCenterOfTile(ITileFieldPlayer player)
...
+
+        public override void UpdateGraphicsToMatchSurroundings()
+        {
+            int x = Location.X;
+            TileField tileField = Parent as TileField;
+
+            if (x == 0 || x == tileField.Columns)
+            {
+                sprite = new SpriteSheet("verticalEnd");
+            }
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/TilesTypes/EndTile.cs:27]: Add an else to make it a horizontal exit. Don't rely on earlier states of the tile.

File: MeesGame/Gameplay/Generator/MazeGeneratorMain.cs [hunk 5]
```diff
@@ -117,8 +118,8 @@ private void MazeGenMain()
             // Add the startTile.
             tiles.Add(new StartTile(0, "playerstart"), start.X, start.Y);
 
-            // Add an end tile on the best exit.
-            tiles.Add(new EndTile(), bestExit.X, bestExit.Y);
+            // Add an end tile on the edge next to the best exit.
+            tiles.Add(new EndTile(), FindEdgeTile(bestExit).X, FindEdgeTile(bestExit).Y);
```
Comment (jensj12) [inline MeesGame/Gameplay/Generator/MazeGeneratorMain.cs]: Only call findedgetile once

File: MeesGame/Gameplay/Generator/MazeGeneratorInit.cs [hunk 6]
```diff
@@ -154,7 +154,7 @@ private void InitialiseExitSearch()
         {
             // The exit will be placed at the hardest to reach tile, so keep an array for that.
             exitScore = new int[numRows, numCols];
-            bestExit = start;
+            bestExit = new Point(0, 1);
```
Comment (jensj12) [inline MeesGame/Gameplay/Generator/MazeGeneratorInit.cs]: Why this change? There's an error handler in case the best exit remains the start and the start isn't on the edge. Using magic values always comes with a risk.

**Review by jensj12 — CHANGES_REQUESTED** [review]

**Comment by jensj12** [comment]
Where exactly does ugtms dereference a null reference?

**Comment by jorisBarkema** [comment]
In the foreach loop in tilefield.updategraphicstomatchsurroundings().

Only thing different from master is that I added the UGTMS override in endtile though, and when I removed that the error was still there.

**Comment by jensj12** [comment]
'Exactly' means the exact instuction and the line of code. If you know where it crashed, you which variable wasn't initialised. Then you can track down where it should have been initialised and fix the crash. In this case, tile is null, which means not every location in the tilegrid has a tile. The callstack tells us that the initialisation failed in the constructor of editorlevel (as it called ugtms from there). It tries to place floortiles on the edge, which isn't allowed.

**Comment by jorisBarkema** [comment]
error has been fixed

**Comment by jensj12** [comment]
OnEdgeOfTileField doesn't work correctly on the right and lower side of the maze. Visible when starting the editor. Also, those three functions contain more code than necessary. Checking if a bool is true and then returning true (and false otherwise) isn't needed.

**Comment by jorisBarkema** [comment]
Oops, silly mistake, fixed it.

**Comment by LisanneW** [comment]
Starting the level editor doesn't immediatly call ugtms() on all tiles. The orientation of the exit is incorrect on the right side of the maze.

**Comment by jorisBarkema** [comment]
should be good now

**Review by LisanneW — APPROVED** [review]

**Review by jensj12 — APPROVED** [review]
